### PR TITLE
A streaming functor on top of a block-based algorithm type class

### PR DESCRIPTION
**Pull request by msprotz**
This pull request introduces in `code/streaming` a facility for automatically generating a streaming API (i.e. a byte-by-byte APIs with an internal buffer) given a block-based API.

To that end, I capture in `Hacl.Streaming.Interface.fsti` the type class of block-based operations. This type class (encoded as an inductive with a single constructor) captures:
- a notion of low-level state, footprint, invariant and framing lemmas (Low* code)
- a notion of a proof-level pure state, and a `v` function from low-level state to proof-level state
- a notion of a high-level block-based specification (spec)
- a notion of a high-level one-shot specification (spec)
- a set of desired lemmas for the high-level specification (associativity of the fold function, and equivalence of block-by-block and one-shot specifications)
- a set of low-level operations implementing the high-level specification.

Then, given an instance of a type class, `Hacl.Streaming.Functor.fsti` will create a low-level streaming API for the given instance. Importantly, once instantiated, the streaming functor:
- is equipped with an abstract interface, making it easy to re-seal the abstraction
- monomorphizes entirely to low-level code, using some advanced tricks to bypass higher-kinded types that don't extract to ML polymorphism
- uses the C abstract struct facility to ensure a pleasant-looking .h file
- takes an (optional) index for (optional) run-time agility

Two instances of the type class are provided as examples:
- EverCrypt.Hash.Incremental is converted to use this new facility
  - it keeps a run-time index for agility
  - it leverages the abstraction principles provided by the functor to keep the exact same abstract interface that was there before
- Hacl.Streaming.SHA2_256 is a new example
  - there no run-time agility (the index is `unit`)
  - I did not bother adding an .fsti

Most of the code for the streaming functor comes from the earlier specialized version in EverCrypt.Hash.Incremental. As part of this pull request, several proofs have been converted to not rely on non-linear arithmetic, and several others have been fleshed out to be more robust. Anecdotally, proof performance is much better for the streaming functor since it operates in a small context that does not include, say, hashes, since all the required operations are codified in Hacl.Streaming.Interface

Getting this to extract properly in Low* is delicate. In essence, `Functor.state_s` is not ML-polymorphic since it takes a type class and an index, both of which get erased to `unit`, resulting in a lot of Obj.magic's and generally invalid Low* code. To workaround this, I introduce an extra type parameter `t` at `Type0` that is known statically to be no other than `c.state i`. The extra indirection, however, means that instead of having:

```
type streaming_state #index (c: my_type_class index) (i: index) = {
  block_state: c.state i; // erases to unit
}
```

I now have something that looks like a regular parameterized type definition in ML, and hence extracts to a parameterized data type in the ML AST then gets monomorphized properly by KreMLin.

```
type streaming_state #index (c: my_type_class index) (i: index) (t: Type0 { t == c.state i }) = {
  block_state: t; // extracts to type 'a streaming_state = { block_state: 'a }
}
```

This extra type parameter appears in every low-level function that needs to extract with prenex polymorphism for the sake of, once more, generating valid Low* code that can undergo type-based monomorphization. The added verbosity is not too bad.

Now that this is done, we know for sure that we can generate incremental versions of all the hash algorithms and of course blake2 once it makes its way into the agile hash API.

My hope is that @prosecco or @polubelova can take a look at this and try to create a couple more instances of this type class with algorithms from other families. I'm particularly interested in Poly105.

Note that this is all operating on Lib.IntTypes so there should be no problem interacting with lib-heavy code, apart from the usual type casts for buffers.

### CODE REVIEW

**Comment by msprotz** [comment]
The second version of this pull request is now up for review. Compared to the previous version, a huge amount of things are improved.
- The update_multi based specifications are moved to a generic spec file, Spec.UpdateMulti, that is parameterized over an update function; this allows moving proofs to a central location and sharing definitions between hashes and streaming, which both are done in terms of update-multi.
- Unfortunately, Lib does not use the update-multi abstraction, but rather the repeat-blocks abstraction. I introduce Hacl.Streaming.Lemmas which does a proof of spec equivalence between update-multi and repeat-blocks based functions.
- I refactored the block interface to separate the `stateful` type class (a piece of state that obeys standard framing principles) from the main block-based API.
- I also refactored the block interface to take *two* instances of stateful: one for the internal state, and one for the type of keys.
- Next, I generalize the streaming functor to keep track of the key at run-time. Algorithms that are block-based (like hashes) but do not need a key at all can pass a dummy instance of `stateful` for their key, which being just a unit, gets eliminated by kremlin.
- The streaming functor has two key management modes:
  - in `Erased`, the key is taken by the `block.init` function at run-time but then retained as an erased value and passed ghostly to `block.finish` -- this is useful for, say, blake2 (*NOT* in this pull request)
  - in `Runtime`, the key is received by both `block.init` and `block.finish`; the streaming functor retains in its internal state the key at run-time so as to pass it to `finish`-- this is useful for, say, Poly1305

In order to demonstrate the usefulness of all of that, I add a new instance of the type class for poly1305/32, wherein I use the various spec-equivalence lemmas to prove that poly1305 is an instance of this type class.

Owing to the fancy indexing and humongous amounts of inline_for_extraction, you can see in the generated code (this diff is clean!) that streaming-sha256 has three fields for its state, but streaming-poly1305 has one more, for keeping the key at run-time.

Being smart about indexing makes my life difficult, but ensures that there are no superfluous arguments in the generated code. In particular, the `Erased` mode will ensure that blake2 does not waste a word of space keeping a key at run-time that will not actually be needed upon calling `finish`.

I would like to merge this soon, since I believe the spec-equivalence lemmas are going to be very useful for @R1kM who will need those for his proof of blake2-is-an-agile-hash.

File: dist/gcc-compatible/Hacl_Streaming_Poly1305_32.c [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+
+#include "Hacl_Streaming_Poly1305_32.h"
+
+typedef struct Hacl_Streaming_Functor_state_s___uint64_t___uint8_t__s
+{
+  uint64_t *block_state;
+  uint8_t *buf;
+  uint64_t total_len;
+  uint8_t *maybe_key;
```
Comment (R1kM) [inline dist/gcc-compatible/Hacl_Streaming_Poly1305_32.c:32]: This name is not ideal, cf earlier comment in Streaming.Functor

File: specs/Spec.Poly1305.fst [hunk 2]
```diff
@@ -75,3 +75,4 @@ let poly1305_mac (msg:bytes) (k:key) : Tot tag =
   let acc, r = poly1305_init k in
   let acc = poly1305_update msg acc r in
   poly1305_finish k acc
+
```
Comment (R1kM) [inline specs/Spec.Poly1305.fst:78]: Extra line?

File: code/streaming/Hacl.Streaming.Functor.fst [hunk 3]
```diff
@@ -0,0 +1,1043 @@
...
+=
+| State:
+    block_state: t ->
+    buf: B.buffer Lib.IntTypes.uint8 { B.len buf = c.block_len i } ->
+    total_len: UInt64.t ->
+    seen: G.erased (S.seq uint8) ->
+    // Stupid name conflict on overloaded projectors leads to inscrutable
+    // interleaving errors. Need a field name that does not conflict with the
+    // one in Hacl.Streaming.Interface. Sigh!!
+    maybe_key: t' ->
```
[the remark this answers is not in the document]
Comment (msprotz) [reply code/streaming/Hacl.Streaming.Functor.fst:44]: fixed

File: code/streaming/Hacl.Streaming.Functor.fsti [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+/// Provided an instance of the type class, this creates a streaming API on top.
+/// This means that every function in this module takes two extra arguments
+/// compared to the previous streaming module specialized for hashes: the type
+/// of the index, and a type class for that index. Then, as usual, a given value
+/// for the index as a parameter.
+///
+/// This streaming API only allocates its internal state on the heap, no support
+/// for allocation on the stack via StackInline.
+
+#set-options "--max_fuel 0 --max_ifuel 0"
```
[the remark this answers is not in the document]
Comment (msprotz) [reply code/streaming/Hacl.Streaming.Functor.fsti:12]: fixed, thanks

File: code/streaming/Hacl.Streaming.Functor.fsti [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+module U64 = FStar.UInt64
+module I = Hacl.Streaming.Interface
+
+open Hacl.Streaming.Interface
+open FStar.HyperStack.ST
+open LowStar.BufferOps
+open FStar.Mul
+
+inline_for_extraction noextract
+let uint8 = Lib.IntTypes.uint8
```
[the remark this answers is not in the document]
Comment (msprotz) [reply code/streaming/Hacl.Streaming.Functor.fsti:29]: trying to limit name pollution... I'm very weary of using `open` excessively... stylistic thing

File: code/streaming/Hacl.Streaming.Interface.fsti [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+  s: (index -> Type0) ->
+
+  // Astract footprint.
+  footprint: (#i:index -> h:HS.mem -> s:s i -> GTot B.loc) ->
+  freeable: (#i:index -> h:HS.mem -> p:s i -> Type) ->
+  invariant: (#i:index -> h:HS.mem -> s:s i -> Type) ->
+
+  // A pure representation of a s
+  t: (index -> Type0) ->
+  v: (i:index -> h:HS.mem -> s:s i -> GTot (t i)) ->
```
[the remark this answers is not in the document]
Comment (msprotz) [reply code/streaming/Hacl.Streaming.Interface.fsti:60]: The typical implementation of `v` is `B.as_seq` which returns `GTot`, so I'm following that convention. It's easy to do a wrapper if someone needs it via `G.hide` -- I'm also weary of the extra SMT burden to unwrap the `erased` / unpredictability of auto-reveal insertion

**Comment by msprotz** [comment]
The ineffiency that was initially present in poly1305 (superfluous calls to init) has now been removed thanks to @polubelova's lemma.

**Comment by msprotz** [comment]
I moved everything related to UpdateMulti to `Lib.UpdateMulti.Lemmas`. I left `update_full` in `Lib.UpdateMulti` because it's a function, not a lemma, so I thought it didn't belong to the `Lemmas` suffix.

I tried moving  `Hacl.Streaming.Spec` to `lib/` but it relies heavily on `Hacl.Streaming.Interface.fsti` and then that would've introduced a dependency edge from `lib` to `code` which we usually try to avoid. So, I thought we could revisit that later.

I authored two C API tests in `tests/`. This API is clearly much nicer to use than the previous block-based ones.

**Comment by msprotz** [comment]
Thanks to @R1kM and @polubelova for the review! I merged after getting an all-green on CI (see last two commits).

**Pull request merged**